### DISCUSSION TITLE
blockscout's design architecture

### DISCUSSION THREAD

**Discussion by formingform** [blockscout/blockscout#8874] [Q&A]
Can someone roughly describe the system design architecture of the entire blockscout project, especially the backend services, such as what services are available, their respective functions and dependencies between a, data flow, etc.? As a developer, I'm trying to do some custom development for my own private chain and need to understand this before I can get my hands dirty.

**Comment by Mojmir-R** [comment]
As your question is pretty complex, I would advise you to first check our docs, best starting point is this page: https://docs.blockscout.com/for-developers/information-and-settings You should be able to find answers to all your questions there. If not, feel free to follow up and ask here.

**Comment by formingform** [comment]
First of all, thank you for your answer.
Then back to my question. What I probably know now is that the whole blockscout project consists of 3 sub-projects: blockscout, frontend, blockscout-rs. the role of frontend is self-explanatory, but for the blockscout, blockscout-rs sub-projects, what are their roles? What are the services? How do these services collaborate? Take for example, I have a POS protocol private chain with a set of smart contracts deployed on it that manages all the verifier's basic information, stkaing information, and information about the rewards earned, etc. I want to add some front-end pages to display all these information. In order to achieve this goal, usually, besides the front-end pages, the back-end needs to provide some additional functions and services, such as fetching these information from the chain and doing storage, calculation, statistics, re-storage, and finally providing some RPC interfaces to the front-end. So my question is, in such a case, then how should I come to do customized development based on blockscout? Or rather, in each sub-project, what should I do?

**Reply by vbaranov** [reply]
blockscout repository contains the core of the explorer, written on Elixir, which contains main indexers (realtime and catchup) and a lot of secondary services (fetching token instances, rewards, internal transactions, on-demand coin/token balances etc.). All of the fetched data are stored in the Postgres DB. Apart from that it contains API (Restful, Etherscan-like, node's JSON RPC-like, GraphQL) and an old embedded UI.
blockscout-rs is the repository for microservices written on Rust like eth bytecode DB, smart-contract verification, sol2uml visualization of smart-contracts, signature provider, stats. Some of the services (e.g. stats) use the main Blockscout DB in order to process the data, others are stateless (smart-contract verification, sol2uml visualization, signature provider) or use their own DB (e.g. ETH bytecode DB).

Depending where are you skilled better Rust or Elixir it is up to you where to write additional fetchers from Blockchain.

In order to familiarize yourself with the indexer structure please read https://github.com/blockscout/blockscout/tree/master/apps/indexer#readme.
Blockscout main DB schema: https://github.com/blockscout/blockscout-db-schema

**Comment by formingform** [comment]
Also, does blockscout support POS consensus and corresponding economic system?

**Reply by vbaranov** [reply]
Yes, Blockscout supports POS EVM compatible blockchains. You can check https://eth.blockscout.com/.
